src/mmds/src/data_store.rs

```diff
@@ -89,15 +72,77 @@ impl Mmds {
         self.data_store.to_string()
     }
 
-    /// This function replicates the behavior of the Instance Metadata Service
+    /// Returns the serde::Value in IMDS format plaintext.
+    /// Currently, only JSON objects and strings can be IMDS formatted.
+    ///
+    /// See the docs for detailed description of the IMDS format:
     /// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html
-    /// 1. For a (key, value) pair where the value is a dictionary, it will return all the keys
-    /// in the dictionary.
-    /// 2. For a (key, value) pair where the value is a simple type (bool, string, number),
-    /// it will return the value.
     ///
-    /// When the path is not found, a NotFound error is returned.
-    pub fn get_value(&self, path: String) -> Result<Vec<String>, Error> {
+    /// # Examples
+    ///
+    /// ```json
+    /// {
+    ///     "key1" : {
+    ///         "key11": "value11"
+    ///         "key12": "value12"
+    ///     }
+    ///     "key2" : "value3"
+    ///     "key3" : "value3"
+    /// }
+    ///```
+    ///
+    /// IMDS formatted JSON object:
+    /// ```text
+    /// key1/
+    /// key2
+    /// key3
+    /// ```
+    ///
+    /// JSON string:
+    /// ```json
+    /// "value"
+    /// ```
+    ///
+    /// IMDS formatted string:
+    /// ```text
+    /// value
+    /// ```
+    ///
+    /// If the `serde_json::Value` is not supported, an `UnsupportedValueType` error is returned.
+    fn format_imds(json: &Value) -> Result<String, Error> {
+        // If the `dict` is Value::Null, Error::NotFound is thrown.
+        // If the `dict` is not a dictionary, a Vec with the value corresponding to
+        // the key is returned.
+        match json.as_object() {
+            Some(map) => {
+                let mut ret = Vec::new();
+                // When the object is a map, push all the keys in the Vec.
+                for key in map.keys() {
+                    let mut key = key.clone();
+                    // If the key corresponds to a dictionary, a "/" is appended
+                    // to the key name.
+                    if map[&key].is_object() {
+                        key.push_str("/");
+                    }
+
+                    ret.push(key);
+                }
+                Ok(ret.join("\n"))
+            }
+            None => {
+                // When the object is not a map, return the value.
+                // Support only `Value::String`.
+                match json.as_str() {
+                    Some(str_val) => Ok(str_val.to_string()),
+                    None => Err(Error::UnsupportedValueType),
+                }
+            }
+        }
+    }
+
+    /// Returns the subtree located at path. When the path corresponds to a leaf, it returns the value.
+    /// Returns Error::NotFound when the path is invalid.
+    pub fn get_value(&self, path: String, accept: MediaType) -> Result<String, Error> {
```

Comment:
At this scope, we shouldn't "leak" information about the upper levels. As far as this layer is concerned, it should be able to return values from the data store in a specified format. In this context, `accept` doesn't make sense. We can rename it to `format`.

When looking at separation of concerns, it should also not use `MediaType` because that is the interface used by HTTP. The MMDS data store is not necessarily tied to HTTP and should define its own interface. The formats that the Data Store supports are JSON and IMDS. At a higher level we should make the translation between ApplicationJson -> Json and PlainText -> Imds.

In the data-store module:
```rust
enum OutputFormat {
    Json
    Imds
}

....
pub fn get_value(&self, path: String, format: OutputFormat) -> Result <String, Error> .....
```

In the `mmds/src/lib.rs` (this should eventually be moved some place else where it makes more sense):
```rust
impl Into<OutputFormat> for MediaType {
   fn into (&self) ... {
       match self {
                MediaType::ApplicationJson => OutputFormat::Json,
                MediaType::PlainText => OutputFormat::Imds,
       }
}
```

Reply:
Done.